modules/library/metadata/src/main/java/org/geotools/factory/FactoryRegistry.java

```diff
@@ -152,15 +158,52 @@ public FactoryRegistry(final Class<?>[] categories) {
      * @param categories The categories.
      */
     public FactoryRegistry(final Collection<Class<?>> categories) {
-        super(categories.iterator());
-        for (final Iterator<Class<?>> it=getCategories(); it.hasNext();) {
-            if (needScanForPlugins == null) {
-                needScanForPlugins = new HashSet<Class<?>>();
-            }
-            needScanForPlugins.add(it.next());
+        for (Class<?> category : categories) {
+            needScanForPlugins.add(category);
+            registry.registerCategory(category);
         }
     }
 
+    /*
+     * TODO: discuss how much API can be broken
+     *
+     * Breakage:
+     *
+     *  - not all methods of `ServiceRegistry` API were implemented
+     *
+     * Possible breakage:
+     *
+     *  - in general nomenclature, replace "ServiceProvider" with "Factory"
+     *  - RegisterableService interface can no longer be used because it expects a `ServiceProvider`
+     *  - replace `Filter` with `Predicate<? super T>`
```

Comment:
I thought about this and decided to go with `Predicate`. True, `Filter` is still public but I find it weird to depend on some arbitrary interface instead of one that was intended for it. It also simplified code a little because the `filter` (nor `test`) method no longer has to handle `Object` but can be written to only deal with the intended factory type.